File: ligature-in-memory/src/main/scala/dev/ligature/inmemory/InMemoryLigature.scala
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.inmemory

// import dev.ligature.{Ligature, QueryTx, Triple}
// import scala.collection.mutable.{Set, TreeMap}
// import dev.ligature.DatasetName

// protected case class DatasetStore(var counter: Long = 0, val edges: Set[Triple] = Set())

case class Test(val x: Int)

// final class LigatureInMemory(
//     private val store: TreeMap[DatasetName, DatasetStore] = TreeMap[DatasetName, DatasetStore]()
// ) extends Ligature {
// //  private val store = AtomicCell[IO].of(TreeMap[Dataset, DatasetStore]())

//   /** Returns all Datasets in a Ligature instance. */
//   override def allDatasets(): Iterator[DatasetName] = store.keySet.iterator

//   /** Check if a given Dataset exists. */
//   override def networkExists(network: DatasetName): Boolean = store.contains(network)

//   /** Returns all Datasets in a Ligature instance that start with the given
//     * prefix.
//     */
//   override def matchDatasetsPrefix(prefix: String): Iterator[DatasetName] =
//     store.keys.filter(_.name.startsWith(prefix)).iterator

//   /** Returns all Datasets in a Ligature instance that are in a given range
//     * (inclusive, exclusive].
//     */
//   override def matchDatasetsRange(
//       start: String,
//       end: String
//   ): Iterator[DatasetName] =
//     store.keys.filter(k => k.name >= start && k.name < end).iterator

//   /** Creates a network with the given name. TODO should probably return its own
//     * error type { InvalidDataset, DatasetExists, CouldNotCreateDataset }
//     */
//   override def createDataset(network: DatasetName): Unit =
//     if !this.store.contains(network) then this.store += (network -> DatasetStore())

//   /** Deletes a network with the given name. TODO should probably return its own
//     * error type { InvalidDataset, CouldNotDeleteDataset }
//     */
//   override def deleteDataset(network: DatasetName): Unit =
//     val _ = this.store.remove(network)

//   override def allTriples(network: DatasetName): Iterator[Triple] =
//     this.store.get(network) match
//       case None        => Iterator.empty
//       case Some(value) => value.edges.iterator

//   /** Initializes a QueryTx TODO should probably return its own error type
//     * CouldNotInitializeQueryTx
//     */
//   override def query[T](network: DatasetName)(fn: QueryTx => T): T =
//     store.get(network) match
//       case None => ???
//       case Some(networkStore) =>
//         val tx = InMemoryQueryTx(networkStore)
//         fn(tx)

//   override def addTriples(network: DatasetName, edges: Iterator[Triple]): Unit =
//     this.store.get(network) match {
//       case None =>
//         this.store.addOne((network, DatasetStore(0, Set.from(edges.toSet))))
//       case Some(store) => store.edges.addAll(edges)
//     }
//     // store.get(network) match
//     //   case None => ???
//     //   case Some(networkStore) =>
//     //     edges
//     //       .fold(networkStore) { (networkStore, edge) =>
//     //         ???
//     //         //networkStore.copy(edges = networkStore.edges + edge)
//     //       }
//     //       .map { networkStore =>
//     //         store.updated(network, networkStore)
//     //       }

//   override def removeTriples(network: DatasetName, edges: Iterator[Triple]): Unit =
//     this.store.get(network) match {
//       case None        => ???
//       case Some(store) => store.edges.subtractAll(edges)
//     }
//     // store.get(network) match
//     //   case None => ???
//     //   case Some(networkStore) =>
//     //     edges
//     //       .fold(networkStore) { (networkStore, edge) =>
//     //         networkStore.copy(edges = networkStore.edges - edge)
//     //       }
//     //       .map { networkStore =>
//     //         store.updated(network, networkStore)
//     //       }

//   override def close(): Unit =
//     ()
// }

File: ligature-in-memory/src/main/kotlin/dev/ligature/inmemory/InMemoryLigature.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.inmemory

import arrow.core.Either
import dev.ligature.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.asFlow
import java.util.*
import java.util.concurrent.locks.ReadWriteLock
import java.util.concurrent.locks.ReentrantReadWriteLock

data class DatasetStore(val statements: Set<Statement>)

/** A trait that all Ligature implementations implement. */
class InMemoryLigature(): Ligature {
    private var store = TreeMap<Dataset, DatasetStore>()
    private val lock: ReadWriteLock = ReentrantReadWriteLock()

    /** Returns all Datasets in a Ligature instance. */
    override suspend fun allDatasets(): Flow<Either<LigatureError, Dataset>> {
        val l = lock.readLock()
        return try {
            l.lock()
            store.keys.map { Either.Right(it) }.asFlow()
        } finally  {
            l.unlock()
        }
    }

    /** Check if a given Dataset exists. */
    override suspend fun datasetExists(dataset: Dataset): Either<LigatureError, Boolean> {
        val l = lock.readLock()
        return try {
            l.lock()
            Either.Right(store.contains(dataset))
        } finally {
            l.unlock()
        }
    }

    /** Returns all Datasets in a Ligature instance that start with the given prefix. */
    override suspend fun matchDatasetsPrefix(prefix: String): Flow<Either<LigatureError, Dataset>> {
        val l = lock.readLock()
        return try {
            l.lock()
            store.filter { (key, _) -> key.name.startsWith(prefix) }.keys.map { Either.Right(it) }.asFlow()
        } finally {
            l.unlock()
        }
    }

    /** Returns all Datasets in a Ligature instance that are in a given range (inclusive, exclusive]. */
    override suspend fun matchDatasetsRange(
            start: String,
            end: String,
    ): Flow<Either<LigatureError, Dataset>> {
        val l = lock.readLock()
        return try {
            l.lock()
            store.filter { (key, _) -> key.name >= start && key.name < end }.keys.map { Either.Right(it) }.asFlow()
        } finally {
            l.unlock()
        }
    }

    /** Creates a dataset with the given name.
     * TODO should probably return its own error type { InvalidDataset, DatasetExists, CouldNotCreateDataset } */
    override suspend fun createDataset(dataset: Dataset): Either<LigatureError, Unit> {
        val l = lock.writeLock()
        return try {
            l.lock()
            if (store.contains(dataset)) {
                Either.Right(Unit)
            } else {
                store[dataset] = DatasetStore(setOf())
//                val newStore = store.put(dataset, DatasetStore(0L, setOf()))
//                store = newStore
                Either.Right(Unit)
            }
        } finally {
            l.unlock()
        }
    }

    /** Deletes a dataset with the given name.
     * TODO should probably return its own error type { InvalidDataset, CouldNotDeleteDataset } */
    override suspend fun deleteDataset(dataset: Dataset): Either<LigatureError, Unit> {
        val l = lock.writeLock()
        return try {
            l.lock()
            if (store.contains(dataset)) {
//                val newStore = store - dataset
//                store = newStore
                store.remove(dataset)
                Either.Right(Unit)
            } else {
                Either.Right(Unit)
            }
        } finally {
            l.unlock()
        }
    }

    /** Initializes a QueryTx
     * TODO should probably return its own error type CouldNotInitializeQueryTx */
    override suspend fun <T> query(dataset: Dataset, fn: suspend (QueryTx) -> T): T {
        val l = lock.readLock()
        return try {
            l.lock()
            val datasetStore: DatasetStore? = store[dataset]
            if (datasetStore != null) {
                val queryTx = InMemoryQueryTx(datasetStore)
                fn(queryTx)
            } else {
                TODO()
            }
        } finally {
            l.unlock()
        }
    }

    /** Initializes a WriteTx
     * TODO should probably return its own error type CouldNotInitializeWriteTx */
    override suspend fun <T> write(dataset: Dataset, fn: suspend (WriteTx) -> T): T {
        val l = lock.writeLock()
        return try {
            l.lock()
            val datasetStore: DatasetStore? = store[dataset]?.copy()
            if (datasetStore != null) {
                val writeTx = InMemoryWriteTx(datasetStore)
                val res = fn(writeTx)
                if (!writeTx.isCanceled()) {
                    store[dataset] = writeTx.newDatasetStore()
                    res
                } else {
                    res
                }
            } else {
                TODO()
            }
        } finally {
            l.unlock()
        }
    }
}
